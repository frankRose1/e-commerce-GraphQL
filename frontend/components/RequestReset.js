import React, { Component } from 'react';
import gql from 'graphql-tag';
import {Mutation} from 'react-apollo';
import Error from './ErrorMessage';
import Form from './styles/Form';

const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!){
    requestReset(email: $email){
      message
    }
  }
`;

class RequestReset extends Component {
  state = {
    email: ''
  }

  handleInputChange = e => {
    const {value, name} = e.target;
    this.setState({ [name]: value });
  }

  render() {
    return (
      <Mutation
        mutation={REQUEST_RESET_MUTATION}
        variables={this.state}>
        {(requestReset, {error, loading, called}) => (
          <Form
            data-test="request-reset"
            method="post" 
            onSubmit={async e => {
            e.preventDefault();
            //could aslo set the success messagge to state and display a pop up when this resolves
            await requestReset();
            this.setState({ email: '' })
          }}>
          {!error && !loading && called && (<p>Success! Check your email for a reset link!</p>)}
          <fieldset disabled={loading} aria-busy={loading}>
            <Error error={error}/>
            <h2>Request Password Reset</h2>

            <label htmlFor="email">
              Email Address
              <input 
                type="email" 
                id="email"
                name="email"
                placeholder="Email"
                value={this.state.email}
                onChange={this.handleInputChange}
                required/>
            </label>

            <button 
              type="submit"
              disabled={loading}>
              Request{loading ? 'ing': ''} Reset!
            </button>
          </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default RequestReset;
export {REQUEST_RESET_MUTATION};