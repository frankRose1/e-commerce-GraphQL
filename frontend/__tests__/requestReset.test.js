import {mount} from 'enzyme';
import {MockedProvider} from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import RequestReset, {REQUEST_RESET_MUTATION} from '../components/RequestReset'

const mocks = [
  {
    request: { query: REQUEST_RESET_MUTATION, variables: { email: 'test@test.com' }},
    result: {
      data: { requestReset: { message: 'Success!', __typename: 'SuccessMessage'} }
    }
  }
];

describe('<RequestReset />', () => {

  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );
    const form = wrapper.find('form[data-test="request-reset"]');
    expect(toJSON(form)).toMatchSnapshot();
  });
  
  it('calls the mutation', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );
    //simulate typing in to the email input, passing in the expected event.target
    wrapper
      .find('input')
      .simulate('change', {target: { name: 'email', value: 'test@test.com' } });
    //submit the form
    wrapper.find('form').simulate('submit');
    await wait(50);
    wrapper.update();
    expect(wrapper.find('p').text()).toContain("Success! Check your email for a reset link!");
  });

})