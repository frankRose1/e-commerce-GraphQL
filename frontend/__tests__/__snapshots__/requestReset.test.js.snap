// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-14226r0-0 kzFose"
  data-test="request-reset"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <DisplayError
      error={Object {}}
    />
    <h2>
      Request Password Reset
    </h2>
    <label
      htmlFor="email"
    >
      Email Address
      <input
        id="email"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        required={true}
        type="email"
        value=""
      />
    </label>
    <button
      disabled={false}
      type="submit"
    >
      Request
       Reset!
    </button>
  </fieldset>
</form>
`;
